save-cloud-common/src/commonMain/kotlin/com/saveourtool/save/cvsscalculator/ScoreCalculator.kt

```diff
@@ -0,0 +1,115 @@
+@file:Suppress("FILE_NAME_MATCH_CLASS")
+
+package com.saveourtool.save.cvsscalculator
+
+import com.saveourtool.save.cvsscalculator.CvssMetrics.Companion.ATTACK_COMPLEXITY
+import com.saveourtool.save.cvsscalculator.CvssMetrics.Companion.ATTACK_VECTOR
+import com.saveourtool.save.cvsscalculator.CvssMetrics.Companion.AVAILABILITY
+import com.saveourtool.save.cvsscalculator.CvssMetrics.Companion.CONFIDENTIALITY
+import com.saveourtool.save.cvsscalculator.CvssMetrics.Companion.CVSS_VERSION
+import com.saveourtool.save.cvsscalculator.CvssMetrics.Companion.INTEGRITY
+import com.saveourtool.save.cvsscalculator.CvssMetrics.Companion.PRIVILEGES_REQUIRED
+import com.saveourtool.save.cvsscalculator.CvssMetrics.Companion.SCOPE
+import com.saveourtool.save.cvsscalculator.CvssMetrics.Companion.USER_INTERACTION
+import kotlin.math.min
+import kotlin.math.pow
+import kotlin.math.roundToInt
+
+@Suppress("UtilityClassWithPublicConstructor")
+private class CvssMetrics {
+    companion object {
+        const val ATTACK_COMPLEXITY = "/AC:"
+        const val ATTACK_VECTOR = "/AV:"
+        const val AVAILABILITY = "/A:"
+        const val CONFIDENTIALITY = "/C:"
+        const val CVSS_VERSION = "CVSS:"
+        const val INTEGRITY = "/I:"
+        const val PRIVILEGES_REQUIRED = "/PR:"
+        const val SCOPE = "/S:"
+        const val USER_INTERACTION = "/UI:"
+    }
+}
+
+@Suppress(
+    "UnsafeCallOnNullableType",
+)
+private fun String.parsingVector(): BaseMetrics {
+    val values = this.toMap()
+    return BaseMetrics(
+        version = values[CVSS_VERSION]!!.toFloat(),
+        attackVector = AttackVectorType.values().find { it.value == values[ATTACK_VECTOR] }!!,
+        attackComplexity = AttackComplexityType.values().findOrElseThrow(ATTACK_COMPLEXITY) { it.value == values[ATTACK_COMPLEXITY] },
+        privilegeRequired = PrivilegesRequiredType.values().findOrElseThrow(PRIVILEGES_REQUIRED) { it.value == values[PRIVILEGES_REQUIRED] },
+        userInteraction = UserInteractionType.values().findOrElseThrow(USER_INTERACTION) { it.value == values[USER_INTERACTION] },
+        scopeMetric = ScopeType.values().findOrElseThrow(SCOPE) { it.value == values[SCOPE] },
+        confidentiality = CiaType.values().findOrElseThrow(CONFIDENTIALITY) { it.value == values[CONFIDENTIALITY] },
+        integrity = CiaType.values().findOrElseThrow(INTEGRITY) { it.value == values[INTEGRITY] },
+        availability = CiaType.values().findOrElseThrow(AVAILABILITY) { it.value == values[AVAILABILITY] },
+    )
+}
+
+private fun <T> Array<out T>.findOrElseThrow(vector: String, predicate: (T) -> Boolean): T = this.find(predicate) ?: throw IllegalArgumentException("No such value for $vector.")
```

Comment:
as a suggestion, it can be rewritten to:

```suggestion
private inline fun <reified T: Enum<T>> Map<String, String>.findOrElseThrow(vector: String, getValue: (T) -> String): T =
    get(vector)?.let { value -> enumValues<T>().find { getValue(it) == value } }
        ?: throw IllegalArgumentException("No such value for $vector.")
```

And then calling can be changed to:
```kotlin
availability = values.findOrElseThrow(AVAILABILITY, CiaType::value),
```

Reply:
Changed it